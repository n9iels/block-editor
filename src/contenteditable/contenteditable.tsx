import * as React from "react";

export interface ContentEditableProps {
    disabled?: boolean;
    html?: string;
    onBlur?: () => void;
    onFocus?: () => void;
    onSelect?: (selectedText: string, rect: ClientRect) => void;
    onSelectStop?: () => void;
    onInput: (value: string) => void;
}

export interface ContentEditableState { selection: Selection }

export class ContentEditable extends React.Component<ContentEditableProps, ContentEditableState> {
    private elementRef: HTMLDivElement;

    constructor(props: ContentEditableProps) {
        super(props);
        this.state = { selection: null };
    }

    componentDidMount() {
        document.execCommand("defaultParagraphSeparator", false, "p");
    }

    shouldComponentUpdate(nextProps: ContentEditableProps) {
        if (nextProps.html !== this.props.html) {
            return true;
        }

        return this.props.disabled !== nextProps.disabled;
    }

    componentDidUpdate(nextProps: ContentEditableProps) {
        // The editable div is not managed by react, thus we have to update the content manualy
        if (this.props.html !== nextProps.html) {
            this.elementRef.innerHTML = this.props.html;

            this.onInput();
        }
    }

    onInput() {
        this.props.onInput(this.elementRef.innerHTML);
    }

    onSelect() {
        const selection = window.getSelection();
        const range = selection.getRangeAt(0);

        if (!selection || !range) {
            return;
        }

        if (this.state.selection && selection.type === "Caret") {
            this.setState({ selection: null }, () => this.props.onSelectStop && this.props.onSelectStop());
        }

        if (selection.type === "Range" && range && range.startOffset !== range.endOffset) {
            const selectedText = selection.toString();
            const rangeclientRect = range.getClientRects().item(0);

            this.setState({ selection: selection }, () => this.props.onSelect && this.props.onSelect(selectedText, rangeclientRect));
        }
    }

    render() {
        return React.createElement(
            "div",
            {
                onInput: () => this.onInput(),
                onBlur: () => this.props.onBlur(),
                onFocus: () => this.props.onFocus(),
                onSelect: () => this.onSelect(),
                ref: (element: HTMLDivElement) => { this.elementRef = element; },
                suppressContentEditableWarning: true,
                contentEditable: !this.props.disabled,
                dangerouslySetInnerHTML: { __html: this.props.html ? this.props.html : "<p><br></p>" }
            }
        );
    }
}
