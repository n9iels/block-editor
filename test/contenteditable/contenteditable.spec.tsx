import * as React from "react";
import * as Renderer from "react-test-renderer";
import { ContentEditable, ContentEditableProps, ContentEditableState } from "../../src/contenteditable/contenteditable";
import { configure, mount, ReactWrapper } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("contenteditable", () => {
    let mockExecCommand: jest.Mock<boolean>;
    let component: ReactWrapper<ContentEditableProps, ContentEditableState, ContentEditable>;

    beforeEach(() => {
        // Jest does not support document.execCommand, so we have to mock it
        // eslint-disable-next-line
        mockExecCommand = jest.fn((command, showUI, value) => true);
        document.execCommand = mockExecCommand;

        component = mount(<ContentEditable onInput={jest.fn()} />);
    });

    it("should perform initial setup", () => {
        expect(Renderer.create(<ContentEditable onInput={jest.fn()} />).toJSON()).toMatchSnapshot();
        expect(mockExecCommand).toBeCalledWith("defaultParagraphSeparator", false, "p");
    });

    it("should rerender when 'html' input property is changed", () => {
        component.setProps({ html: "<p>content</p>" });
        expect(component.find("div[contentEditable=true]").html()).toBe("<div contenteditable=\"true\"><p>content</p></div>");

        component.setProps({ html: "<p>content update</p>" });
        expect(component.find("div[contentEditable=true]").html()).toBe("<div contenteditable=\"true\"><p>content update</p></div>");
    });

    it("should disable the editable div when property 'disabled' is true", () => {
        component.setProps({ disabled: true });

        expect(component.find("div[contentEditable=true]").exists()).toBeFalsy();
    });

    it("should clear the selection state and call the prop 'onSelectStop' when the selection is cleared", () => {
        window.getSelection = jest.fn().mockReturnValue({ type: "Caret", getRangeAt: jest.fn });
        const onSelectStopMock = jest.fn();
        const onSelectMock = jest.fn();

        component.setProps({ onSelect: onSelectMock, onSelectStop: onSelectStopMock });
        component.setState({ selection: {} as Selection });
        component.instance().onSelect();

        expect(component.state("selection")).toBe(null);
        expect(onSelectStopMock).toBeCalledTimes(1);
        expect(onSelectMock).toBeCalledTimes(0);
    });

    it("should set the state 'selection' and call the prop 'onSelect' when a selection is made", () => {
        const onSelectStopMock = jest.fn();
        const onSelectMock = jest.fn();
        const clientRectMock = jest.fn().mockReturnValue({ item: jest.fn().mockReturnValue({ left: 99 }) });
        const getRangeAtMock = jest.fn()
            .mockReturnValue({ getClientRects: clientRectMock, startOffset: 0, endOffset: 1 });
        window.getSelection = jest.fn().mockReturnValue({ type: "Range", getRangeAt: getRangeAtMock, toString: () => "selected_text" });

        component.setProps({ onSelect: onSelectMock, onSelectStop: onSelectStopMock });
        component.setState({ selection: null });
        component.instance().onSelect();

        expect(onSelectMock).toHaveBeenCalledWith("selected_text", { left: 99 });
        expect(onSelectStopMock).toHaveBeenCalledTimes(0);
    });
});
