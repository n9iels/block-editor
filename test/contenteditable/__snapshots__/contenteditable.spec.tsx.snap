// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contenteditable should perform initial setup 1`] = `
<div
  contentEditable={true}
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p><br></p>",
    }
  }
  onBlur={[Function]}
  onFocus={[Function]}
  onInput={[Function]}
  onSelect={[Function]}
  suppressContentEditableWarning={true}
/>
`;
